packages/space-opera/src/components/materials_panel/materials_panel.ts

```diff
@@ -256,29 +318,149 @@ export class MaterialPanel extends ConnectedLitElement {
+        if (selectedVariantName != null) {
+          this.onSelectVariant(selectedVariantName);
+        }
+      }
+    }}">${
         getModelViewer().availableVariants.map(
-            (name, id) =>
-                html`<paper-item value="${name}">(${id}) ${name}</paper-item>`)}
-      </me-dropdown>
+            (name, id) => html`<paper-item value="${name}">
+            (${id}) ${name}</paper-item>
+            `)}
+        </me-dropdown>
+        <mwc-icon-button icon="create"
+        @click="${() => {
+      this.editVariantNameDialog.textFieldValue = '';
+      this.editVariantNameDialog.placeholder = this.selectedVariant!;
+      this.editVariantNameDialog.open = true;
+    }}"
+        value="${this.selectedVariant}">
+        </mwc-icon-button>
+        </div>
+        <mwc-button
+          label="Create Variant"
+          id="create-variant"
+          @click="${() => {
+      this.createVariantNameDialog.open = true;
+    }}"></mwc-button>
+        </div>
     </me-expandable-tab>
     `;
   }
 
+  validateInput(textField: TextField): boolean {
+    textField.validityTransform =
+        (value: string, nativeValidity: ValidityState) => {
+          // Validates length.
+          if (value.length < 1) {
+            textField.validationMessage = `Invalid input.`;
+            return {valid: false} as ValidityState;
+          }
+
+          // Verifies name is unique.
+          if (getModelViewer().availableVariants.find((existingNames) => {
+                return existingNames === value;
+              })) {
+            textField.validationMessage = `The name ${value} already exists.`;
+            return {valid: false};
+          }
+
+          return nativeValidity;
+        };
+
+    return textField.reportValidity();
+  }
+
+  isValidInput(_value: string): {valid: boolean, validationMessage: string} {
+    // Validates length.
+    if (_value.length < 1) {
+      return {valid: false, validationMessage: `Invalid input.`};
+    }
+
+    // Verifies name is unique.
+    if (getModelViewer().availableVariants.find((existingNames) => {
+          return existingNames === _value;
+        })) {
+      return {
+        valid: false,
+        validationMessage: `The name ${_value} already exists.`
+      };
+    }
+
+    return {valid: true, validationMessage: ''};
+  }
+
+  renderEditVariantDialog() {
+    return html`
+      <input-dialog id="edit-variant-name" modal="true"
+        placeholder="Enter Variant Name">
+      </input-dialog>
+    `;
+  }
+
+  renderCreateVariantDialog() {
+    return html`
+      <input-dialog id="create-variant-name" modal="true"
+        placeholder="Enter Variant Name">
+      </input-dialog>
+    `;
+  }
+
+  renderEditMaterialNameDialog() {
+    return html`
+      <input-dialog id="edit-material-name" modal="true"
+        placeholder="Enter Material Name">
+      </input-dialog>
+    `;
+  }
+
+  onCreateVariant(newVariantName: string) {
```

Comment:
I have no idea where  this  bug is, so I just picked a random line  for this comment. repro:
1) open the robot
2) create a variant
3) switch back to default variant
4) change a color
5) switch to new variant
note that both variants now have the new color instead of just the default.

Reply:
That sounds like intended behavior, editing the material will create a new material for the variant, since the new variant has no change to the material it just utilizes the existing material, this allows the variant to only override a subset of material based on what is changed. You should see in this case that the actual selected material is the same material for both variants.

Reply:
I'm seeing even stranger behavior now. Maybe just revert the  fix for now so we can  merge this and  then we can fix the bug. Unless you see a quick resolution.